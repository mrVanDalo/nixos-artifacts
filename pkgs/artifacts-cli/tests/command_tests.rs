use anyhow::Context;
use anyhow::Result;
#[allow(deprecated)]
use insta_cmd::{StdinCommand, assert_cmd_snapshot, get_cargo_bin};
use serial_test::serial;
use std::fs::remove_dir_all;
use std::path::{Path, PathBuf};
use std::process::Command;

fn cli() -> Command {
    Command::new(get_cargo_bin("artifacts"))
}

#[allow(deprecated)]
fn sdtin_cli(stdin: &str) -> StdinCommand {
    let mut cmd = StdinCommand::new(get_cargo_bin("artifacts"), stdin);
    // StdinCommand::env returns &mut Command; we don't need the return value here.
    let _ = cmd
        .env("TMPDIR", "/tmp/artifacts-cli")
        .arg("--log-level=debug");
    cmd
}

fn project_root() -> PathBuf {
    // tests run with CWD at crate root, but compute robustly
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
}

/// Helper to manage a fixed temp directory for deterministic snapshots.
/// It ensures the directory exists before a test, and after the test
/// verifies it is empty; if not, the test fails, but the directory is
/// deleted in any case.
struct TempTestEnv {
    path: PathBuf,
}

impl Drop for TempTestEnv {
    fn drop(&mut self) {
        // Best-effort cleanup; ignore errors so Drop never panics
        let _ = remove_dir_all(&self.path);
    }
}

#[allow(deprecated)]
impl TempTestEnv {
    fn new() -> Self {
        let path = PathBuf::from("/tmp/artifacts-cli");
        let _ = remove_dir_all(&path);
        std::fs::create_dir_all(&path).expect("failed to create fixed tmp dir");
        TempTestEnv { path }
    }

    fn is_empty_dir(path: &Path) -> std::io::Result<bool> {
        let mut entries = std::fs::read_dir(path)?;
        Ok(entries.next().is_none())
    }

    fn finish(self) -> Result<()> {
        Self::is_empty_dir(&self.path)
            .context("failed to read tmp dir")?
            .then_some(())
            .ok_or_else(|| anyhow::anyhow!("temporary directory not empty before cleanup"))?;
        Ok(())
    }
}

#[test]
#[serial]
fn no_config_scenario() {
    let root = project_root();
    let test_dir = root.join("examples/no_config");

    let env = TempTestEnv::new();

    let mut cmd = sdtin_cli("one\ntwo\n");

    cmd.arg("generate")
        .env(
            "NIXOS_ARTIFACTS_BACKEND_CONFIG",
            &test_dir.join("backend.toml"),
        )
        .arg(test_dir);

    // Verify and cleanup
    env.finish().expect("temp folder not empty at end of test");

    assert_cmd_snapshot!(cmd);
}
#[test]
#[serial]
fn scenario_simple() {
    let root = project_root();
    let test_dir = root.join("examples/scenario_simple");

    let env = TempTestEnv::new();

    let mut cmd = sdtin_cli("one\ntwo\n");

    cmd.arg("generate")
        .env(
            "NIXOS_ARTIFACTS_BACKEND_CONFIG",
            &test_dir.join("backend.toml"),
        )
        .arg(test_dir);

    // Verify and cleanup
    env.finish().expect("temp folder not empty at end of test");

    assert_cmd_snapshot!(cmd);
}

#[test]
#[serial]
fn two_artifacts_scenario() {
    let root = project_root();
    let test_dir = root.join("examples/2_artifacts");

    let env = TempTestEnv::new();

    let mut cmd = sdtin_cli("");

    cmd.arg("generate")
        .env(
            "NIXOS_ARTIFACTS_BACKEND_CONFIG",
            &test_dir.join("backend.toml"),
        )
        .arg(test_dir);

    // Verify and cleanup
    env.finish().expect("temp folder not empty at end of test");
    assert_cmd_snapshot!(cmd);
}

#[test]
#[serial]
fn missing_files_scenario() {
    let root = project_root();
    let test_dir = root.join("examples/missing-files");

    let env = TempTestEnv::new();

    let mut cmd = sdtin_cli("one\ntwo\n");

    cmd.arg("generate")
        .env(
            "NIXOS_ARTIFACTS_BACKEND_CONFIG",
            &test_dir.join("backend.toml"),
        )
        .arg(test_dir);

    // Verify and cleanup
    env.finish().expect("temp folder not empty at end of test");

    assert_cmd_snapshot!(cmd);
}

#[test]
#[serial]
fn wrong_file_type_scenario() {
    let root = project_root();
    let test_dir = root.join("examples/wrong-file-type");

    let env = TempTestEnv::new();

    let mut cmd = sdtin_cli("one\ntwo\n");

    cmd.arg("generate")
        .env(
            "NIXOS_ARTIFACTS_BACKEND_CONFIG",
            &test_dir.join("backend.toml"),
        )
        .arg(test_dir);

    // Verify and cleanup
    env.finish().expect("temp folder not empty at end of test");

    assert_cmd_snapshot!(cmd);
}

#[test]
#[serial]
fn unwanted_files_scenario() {
    let root = project_root();
    let test_dir = root.join("examples/unwanted-files");

    let env = TempTestEnv::new();

    let mut cmd = sdtin_cli("one\ntwo\n");

    cmd.arg("generate")
        .env(
            "NIXOS_ARTIFACTS_BACKEND_CONFIG",
            &test_dir.join("backend.toml"),
        )
        .arg(test_dir);

    // Verify and cleanup
    env.finish().expect("temp folder not empty at end of test");

    assert_cmd_snapshot!(cmd);
}

#[test]
#[serial]
fn missing_generator_scenario() {
    let root = project_root();
    let test_dir = root.join("examples/missing_generator");

    let env = TempTestEnv::new();

    let mut cmd = sdtin_cli("");

    cmd.arg("generate")
        .env(
            "NIXOS_ARTIFACTS_BACKEND_CONFIG",
            &test_dir.join("backend.toml"),
        )
        .arg(test_dir);

    // Verify and cleanup
    env.finish().expect("temp folder not empty at end of test");

    assert_cmd_snapshot!(cmd);
}

#[test]
fn scenario_help() {
    let mut cmd = cli();
    cmd.arg("--help");

    assert_cmd_snapshot!(cmd);
}

#[test]
fn scenario_generator_help() {
    let mut cmd = cli();
    cmd.arg("generate").arg("--help");

    assert_cmd_snapshot!(cmd);
}

#[test]
fn scenario_regenerator_help() {
    let mut cmd = cli();
    cmd.arg("regenerate").arg("--help");

    assert_cmd_snapshot!(cmd);
}

#[test]
#[serial]
fn list_scenarios() {
    let root = project_root();
    let test_dir = root.join("examples/bigger_setup");

    let env = TempTestEnv::new();

    let mut cmd = sdtin_cli("");

    cmd.arg("list")
        .env(
            "NIXOS_ARTIFACTS_BACKEND_CONFIG",
            &test_dir.join("backend.toml"),
        )
        .arg(test_dir);

    // Verify and cleanup
    env.finish().expect("temp folder not empty at end of test");

    assert_cmd_snapshot!(cmd);
}

#[test]
#[serial]
fn bigger_scenarios() {
    let root = project_root();
    let test_dir = root.join("examples/bigger_setup");

    let env = TempTestEnv::new();

    let mut cmd = sdtin_cli("");

    cmd.arg("generate")
        .env(
            "NIXOS_ARTIFACTS_BACKEND_CONFIG",
            &test_dir.join("backend.toml"),
        )
        .arg(test_dir);

    // Verify and cleanup
    env.finish().expect("temp folder not empty at end of test");

    assert_cmd_snapshot!(cmd);
}

#[test]
#[serial]
fn regenerate_all_scenarios() {
    let root = project_root();
    let test_dir = root.join("examples/bigger_setup");

    let env = TempTestEnv::new();

    let mut cmd = sdtin_cli("");

    cmd.arg("regenerate")
        .env(
            "NIXOS_ARTIFACTS_BACKEND_CONFIG",
            &test_dir.join("backend.toml"),
        )
        .arg(test_dir)
        .arg("--all");

    // Verify and cleanup
    env.finish().expect("temp folder not empty at end of test");

    assert_cmd_snapshot!(cmd);
}

#[test]
#[serial]
fn regenerate_machine_scenarios() {
    let root = project_root();
    let test_dir = root.join("examples/bigger_setup");

    let env = TempTestEnv::new();

    let mut cmd = sdtin_cli("");

    cmd.arg("regenerate")
        .env(
            "NIXOS_ARTIFACTS_BACKEND_CONFIG",
            &test_dir.join("backend.toml"),
        )
        .arg(test_dir)
        .arg("--machine=machine-one");

    // Verify and cleanup
    env.finish().expect("temp folder not empty at end of test");

    assert_cmd_snapshot!(cmd);
}

#[test]
#[serial]
fn regenerate_machine_and_artifacts_scenarios() {
    let root = project_root();
    let test_dir = root.join("examples/bigger_setup");

    let env = TempTestEnv::new();

    let mut cmd = sdtin_cli("");

    cmd.arg("regenerate")
        .env(
            "NIXOS_ARTIFACTS_BACKEND_CONFIG",
            &test_dir.join("backend.toml"),
        )
        .arg(test_dir)
        .arg("--machine=machine-one")
        .arg("--artifact=artifact-one");

    // Verify and cleanup
    env.finish().expect("temp folder not empty at end of test");

    assert_cmd_snapshot!(cmd);
}

#[test]
#[serial]
fn regenerate_wrong_machine_scenarios() {
    let root = project_root();
    let test_dir = root.join("examples/bigger_setup");

    let env = TempTestEnv::new();

    let mut cmd = sdtin_cli("");

    cmd.arg("regenerate")
        .env(
            "NIXOS_ARTIFACTS_BACKEND_CONFIG",
            &test_dir.join("backend.toml"),
        )
        .arg(test_dir)
        .arg("--machine=machine-name");

    // Verify and cleanup
    env.finish().expect("temp folder not empty at end of test");

    assert_cmd_snapshot!(cmd);
}

#[test]
#[serial]
fn artifact_name_scenario() {
    let root = project_root();
    let test_dir = root.join("examples/artifact_names");

    let env = TempTestEnv::new();

    let mut cmd = sdtin_cli("");

    cmd.arg("generate")
        .env(
            "NIXOS_ARTIFACTS_BACKEND_CONFIG",
            &test_dir.join("backend.toml"),
        )
        .arg(test_dir)
        .arg("--machine=machine-name");

    // Verify and cleanup
    env.finish().expect("temp folder not empty at end of test");

    assert_cmd_snapshot!(cmd);
}
