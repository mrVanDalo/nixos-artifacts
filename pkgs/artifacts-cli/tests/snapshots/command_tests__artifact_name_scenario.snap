---
source: tests/command_tests.rs
info:
  program: artifacts
  args:
    - "--log-level=debug"
    - generate
    - /home/palo/dev/artifacts/nixos-artifacts/pkgs/artifacts-cli/examples/artifact_names
    - "--machine=machine-name"
  env:
    NIXOS_ARTIFACTS_BACKEND_CONFIG: /home/palo/dev/artifacts/nixos-artifacts/pkgs/artifacts-cli/examples/artifact_names/backend.toml
    TMPDIR: /tmp/artifacts-cli
  stdin: ""
---
success: true
exit_code: 0
----- stdout -----
DEBUG: Running nix build: "/run/current-system/sw/bin/nix" "build" "--impure" "-I" "flake=/home/palo/dev/artifacts/nixos-artifacts/pkgs/artifacts-cli/examples/artifact_names" "--no-link" "--print-out-paths" "--expr" "\nlet\n  system = \"x86_64-linux\";\n  filterAttrs =\n    pred: set:\n    builtins.removeAttrs set (builtins.filter (name: !pred name set.${name}) (builtins.attrNames set));\n  flake = builtins.getFlake (toString <flake>);\n  pkgs = flake.inputs.nixpkgs.legacyPackages.${system};\n  configurations = builtins.attrNames (\n    filterAttrs (\n      machine: configuration: builtins.hasAttr \"artifacts\" configuration.options\n    ) flake.nixosConfigurations\n  );\n  make = map (name: {\n    machine = name;\n    artifacts = flake.nixosConfigurations.${name}.config.artifacts.store;\n    config =\n      if (builtins.hasAttr \"config\" flake.nixosConfigurations.${name}.config.artifacts) then\n        flake.nixosConfigurations.${name}.config.artifacts.config\n      else\n        { };\n  }) configurations;\nin\npkgs.writeText \"test.json\" (builtins.toJSON make)\n"
DEBUG: make config: [{"artifacts":{"artifact.one":{"files":{"fifth secret":{"group":"root","mode":"0400","name":"fifth secret","owner":"root","path":"/run/secrets/fifth secret"},"first.secret":{"group":"root","mode":"0400","name":"first.secret","owner":"root","path":"/run/secrets/first.secret"},"forthSecret":{"group":"root","mode":"0400","name":"forthSecret","owner":"root","path":"/run/secrets/forthSecret"},"second_secret":{"group":"root","mode":"0400","name":"second_secret","owner":"root","path":"/run/secrets/second_secret"},"third-secret":{"group":"root","mode":"0400","name":"third-secret","owner":"root","path":"/run/secrets/third-secret"}},"generator":"/nix/store/wyn91y3kama2cm6pmlf997y49dvb42lh-test_generator.sh","name":"artifact.one","prompts":{},"serialization":"test","shared":false}},"config":{},"machine":"machine-name"}]
[generate]
DEBUG:     files to produce -> 5 files
DEBUG:       - fifth secret => /run/secrets/fifth secret owner=root group=root
DEBUG:       - first.secret => /run/secrets/first.secret owner=root group=root
DEBUG:       - forthSecret => /run/secrets/forthSecret owner=root group=root
DEBUG:       - second_secret => /run/secrets/second_secret owner=root group=root
DEBUG:       - third-secret => /run/secrets/third-secret owner=root group=root
DEBUG:     running check_serialization: env inputs="/tmp/artifacts-cli/inputs-artifact.one" machine="machine-name" artifact="artifact.one" /home/palo/dev/artifacts/nixos-artifacts/pkgs/artifacts-cli/examples/artifact_names/test_check.sh
fifth secret
first.secret
forthSecret
second_secret
third-secret
DEBUG:     check_serialization: failed with status exit status: 1 -> continuing with generation
âš¡ machine-name/artifact.one
DEBUG: bwrap command: bwrap --unshare-all --unshare-user --uid 1000 --gid 1000 --tmpfs / --chdir / --ro-bind /nix/store /nix/store --tmpfs /usr/lib/systemd --proc /proc --dev /dev --bind /tmp/artifacts-cli/prompt-artifact.one /tmp/artifacts-cli/prompt-artifact.one --bind /tmp/artifacts-cli/out-artifact.one /tmp/artifacts-cli/out-artifact.one --ro-bind /nix/store/v36iinbhvz4rn84skvzg7d3qbmr800k4-test_generator.sh/bin /nix/store/v36iinbhvz4rn84skvzg7d3qbmr800k4-test_generator.sh/bin --ro-bind /bin /bin --ro-bind /usr/bin /usr/bin --ro-bind /tmp/artifacts-cli/artifacts-cli-passwd-70cab021bfe35e1e.txt /etc/passwd -- /bin/sh /nix/store/v36iinbhvz4rn84skvzg7d3qbmr800k4-test_generator.sh/bin/test_generator.sh
ðŸ’¾ serialize secrets
=== Content of /tmp/artifacts-cli/out-artifact.one/fifth secret ===
test
=========================
=== Content of /tmp/artifacts-cli/out-artifact.one/first.secret ===
test
=========================
=== Content of /tmp/artifacts-cli/out-artifact.one/forthSecret ===
test
=========================
=== Content of /tmp/artifacts-cli/out-artifact.one/second_secret ===
test
=========================
=== Content of /tmp/artifacts-cli/out-artifact.one/third-secret ===
test
=========================

----- stderr -----
